includes/embeds/class-amp-core-block-handler.php

```diff
@@ -178,4 +180,75 @@ public function ampify_cover_block( $block_content, $block ) {
 		}
 		return $block_content;
 	}
+
+	/**
+	 * Sanitize widgets that are not added via Gutenberg.
+	 *
+	 * @param Document $dom Document.
+	 */
+	public function sanitize_raw_embeds( Document $dom ) {
+		$this->process_categories_widgets( $dom );
+		$this->process_archives_widgets( $dom );
+	}
+
+	/**
+	 * Process "Categories" widgets.
+	 *
+	 * @since 1.5.2
+	 *
+	 * @param Document $dom Document.
+	 */
+	private function process_categories_widgets( Document $dom ) {
+		static $count = 0;
+
+		$widget_containers = $dom->xpath->query( '//*[ contains( @class, "widget_categories" ) ]' );
+		foreach ( $widget_containers as $widget_container ) {
+			if ( ! $widget_container instanceof DOMElement ) {
+				continue;
+			}
+			$select = $dom->xpath->query( './/select[ @name = "cat" ]', $widget_container )->item( 0 );
+			$form   = $widget_container->getElementsByTagName( 'form' )->item( 0 );
+			if ( ! $select instanceof DOMElement || ! $form instanceof DOMElement ) {
+				continue;
+			}
+
+			$count++;
+			$id = sprintf( 'amp-wp-widget-categories-%d', $count );
+			$form->setAttribute( 'id', $id );
+
+			$select->setAttribute( 'on', sprintf( 'change:%s.submit', $id ) );
+			$script = $dom->xpath->query( './/script[ contains( text(), "onCatChange" ) ]', $widget_container )->item( 0 );
+			if ( $script ) {
+				$script->parentNode->removeChild( $script );
+			}
+		}
+	}
+
+	/**
+	 * Process "Archives" widgets.
+	 *
+	 * @since 1.5.2
+	 *
+	 * @param Document $dom Select node retrieved from the widget.
+	 */
+	private function process_archives_widgets( Document $dom ) {
+
+		$widget_containers = $dom->xpath->query( '//*[ contains( @class, "widget_archive" ) ]' );
```

Comment:
This isn't matching widgets that are added via `the_widget()`. For example:

```php
$the_widget_args = array(
		'before_widget' => '<div class="widget">',
		'after_widget'  => '</div>',
		'before_title'  => '<div class="widget-title"><h3>',
		'after_title'   => '</h3></div>',
	);

the_widget( 'WP_Widget_Archives', [ 'title' => 'Archive', 'dropdown' => true ], $the_widget_args );
```

could generate:

```html
<div class="widget">
    <div class="widget-title">
        <h3>Archive</h3>
    </div> <label class="screen-reader-text" for="archives-dropdown--1">Archive</label>
    <select id="archives-dropdown--1" name="archive-dropdown">

        <option value="">Select Month</option>
        <option value="https://wordpressdev.lndo.site/2020/04/"> April 2020 </option>
        <option value="https://wordpressdev.lndo.site/2020/03/"> March 2020 </option>

    </select>

    <script>
        /* <![CDATA[ */
        (function () {
            var dropdown = document.getElementById("archives-dropdown--1");
            function onSelectChange() {
                if (dropdown.options[dropdown.selectedIndex].value !== '') {
                    document.location.href = this.options[this.selectedIndex].value;
                }
            }
            dropdown.onchange = onSelectChange;
        })();
/* ]]> */
    </script>

</div>
```

Reply:
That's true, but if the user is suppressing the class name from appearing then they have bigger problems I think. Themes need that class name to be present in order to style the widgets properly.

Reply:
There are some themes that do override the class name: https://wpdirectory.net/search/01E508QBAE335DVKC95FAV11TD

Reply:
Humm. They do seem to retain the required class name in the class, e.g. `widget_categories`:

![image](https://user-images.githubusercontent.com/134745/78376769-a3f30600-7583-11ea-9b30-afd47971e75d.png)

The exception is Kale which doesn't do it for every widget:

![image](https://user-images.githubusercontent.com/134745/78376858-be2ce400-7583-11ea-884e-4281b2e75b7b.png)

Which may be bug?

Reply:
I suppose this can be hardened by looking for element that matches the query:

```
//select[ @name = 'archive-dropdown' and starts-with( @id, 'archives-dropdown-' ) ]
```

And then the parent element of that is the widget container. That would be invariable.

Reply:
See bdeb953